Data-Structures/HAMT.md

```diff
@@ -1,5 +1,103 @@
-# [WIP] Hash-Array Mapped Trie
+# IPLD HAMT Spec 
 
-This specifies a standardized hash-array mapped trie on IPLD Data Model v1.
+Good reading:
+- https://blog.acolyer.org/2015/11/27/hamt/
+- https://michael.steindorfer.name/publications/oopsla15.pdf
 
-TODO: write this spec.
\ No newline at end of file
+
+The HAMT is a Key Value store implemented with a 256ary hash keyed trie. The
+basic structure in it is a `Node`. The node is as follows:
+
+```go
+type Node struct {
```

Comment:
Could we add an optional seed (defaults to 0)? That way we have room to fix the hashmap DoS attack if necessary.

Reply:
I've been messing with hash algorithm pluggability thinking that (a) different algorithms (and different key lengths) might be optimal in different scenarios, and (b) having the ability to switch it out provides some future-proofing in case of fundamental flaws being discovered in a chosen algorithm. Having space for a seed would also open up space for some keyed algorithms too.
What I can't see (yet) is what kind of use-cases of IPLD are there that attacks against the hash would matter? What's the threat model where this is a concern or is it just a matter of being safe for some as yet unforeseen scenario?

Reply:
The hashmap DoS attack works by attacker inserting keys that hash to the same bucket in a hash map. Something very similar can be done with HAMT by selecting keys to lay in the same branch of the tire.

Reply:
Seed/Nonce works around it by making so the attacker can't simply predict in which branch of the tree will given key lay.

Reply:
I'm not actually sure how adding a seed helps here. It needs to be deterministic, and if its deterministic, then the attacker can know it too and it doesnt make their lives any harder.

I guess forcing a rehash at each layer makes the attack linearly more expensive, but doesnt necessarily prevent attacks.

Reply:
@Kubuxu what I'm really interested in is what kind of use-cases of IPLD exist, or potentially exist, where a hash flood is a practical attack vector? Who are the actors in that kind of scenario? It's obvious in classic uses of maps, like we've had in [node recently](https://nodesource.com/blog/node-js-and-the-hashwick-vulnerability/) but I'm struggling to extrapolate to IPLD.

Reply:
Primary example in my mind is Filecoin. We will be using HAMTs as the primary state tree, and anyone can mutate the state by submitting a transaction.

Reply:
> I'm not actually sure how adding a seed helps here. It needs to be deterministic, and if its deterministic, then the attacker can know it too and it doesnt make their lives any harder.

My suggestion is to make this (optionally) non-deterministic. That way, dapps that notice their HAMTs becoming unbalanced can re-randomize them.

Filecoin probably won't use it (and could assert that the seed is 0) but I'm less worried about filecoin as pulling off this attack would be _expensive_. Although filecoin _could_ still use the seed to rebalance the tree once every N blocks using randomness from the chain).

Reply:
> Who are the actors in that kind of scenario?

Given some kind of consensus driven name registry, an attacker could push a target name deep into the HAMT at the cost of 3 names per layer.

Actually, a better defense might be to:

1. Make the hash function pluggable.
2. Use sha256 where it matters.

An attacker _could_ (by spending a fortune) push a single name ~8-10 layers deep but that's not a very effective attack in a distributed system (assuming graphsync).

Reply:
So, we actually _do_ need to support, e.g., sha256 if we want both security and determinism on systems like Filecoin. Currently, given an N byte insecure hash function, an attacker could create a tree N deep at the target hash, filling the last layer. This could be used to prevent anyone from using a specific key.

Reply:
I have a hard time imagining wanting to use *non-deterministic* balancing in a *distributed* system.  It seems that would produce either massive flapping if actually *used* in a frequently updated dataset, and/or introduces a need for *coordination* where they're previously wasn't one (which is pretty much universally a Bad Thing in a distributed system).  Is there a concrete situation where we can imagine using such a thing, and using it well?

"Use a SHA (or other cryptographic function) when it matters" sounds like a much better approach.  We're already tossing around enough cryptographic functions that it doesn't sound likely to be much of a cost center to introduce another.

Reply:
> I have a hard time imagining wanting to use non-deterministic balancing in a distributed system. It seems that would produce either massive flapping if actually used in a frequently updated dataset, and/or introduces a need for coordination where they're previously wasn't one (which is pretty much universally a Bad Thing in a distributed system). Is there a concrete situation where we can imagine using such a thing, and using it well?

So, the simple use-case is a block-chain where the blockchain determines the seed. Once every N blocks, the hamt would be reseeded automatically.

In general, this will also work for single-writer, multi-reader setups. That's usually the most common case.

---

> "Use a SHA (or other cryptographic function) when it matters" sounds like a much better approach. We're already tossing around enough cryptographic functions that it doesn't sound likely to be much of a cost center to introduce another.

I agree although this still won't be _optimal_. That is, An attacker could pretty easily create a very deep hamt.

---

Basically, I'd prefer to leave room for future improvements now instead of having to introduce them later. However, custom hash functions is _probably_ enough.

Reply:
@Stebalien what youre suggesting would mean rewriting the entire HAMT every epoch. I think thats far worse than a (worst case) 32 deep lookup.

Reply:
I'm saying it could be rebalanced as necessary. However, I agree the better solution is to just use SHA256.

Note: The worst case without using sha256 isn't a 32 deep lookup, it's a full kv list at the max depth preventing further modifications.